node/src/accountant/scanners/pending_payable_scanner/utils.rs

```diff
@@ -1,191 +1,1184 @@
+                )
+            }
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct FailedValidation<RecordStatus> {
+    pub tx_hash: TxHash,
+    pub validation_failure: BlockchainErrorKind,
+    pub current_status: RecordStatus,
+}
+
+impl<RecordStatus> FailedValidation<RecordStatus>
+where
+    RecordStatus: UpdatableValidationStatus,
+{
+    pub fn new(
+        tx_hash: TxHash,
+        validation_failure: BlockchainErrorKind,
+        current_status: RecordStatus,
+    ) -> Self {
+        Self {
+            tx_hash,
+            validation_failure,
+            current_status,
+        }
+    }
+
+    pub fn new_status(&self, clock: &dyn ValidationFailureClock) -> Option<RecordStatus> {
+        self.current_status
+            .update_after_failure(self.validation_failure.clone(), clock)
+    }
+}
+
+pub trait UpdatableValidationStatus {
+    fn update_after_failure(
+        &self,
+        error: BlockchainErrorKind,
+        clock: &dyn ValidationFailureClock,
+    ) -> Option<Self>
+    where
+        Self: Sized;
+}
+
+impl UpdatableValidationStatus for TxStatus {
+    fn update_after_failure(
+        &self,
+        error: BlockchainErrorKind,
+        clock: &dyn ValidationFailureClock,
+    ) -> Option<Self> {
+        match self {
+            TxStatus::Pending(ValidationStatus::Waiting) => Some(TxStatus::Pending(
+                ValidationStatus::Reattempting(PreviousAttempts::new(error, clock)),
+            )),
+            TxStatus::Pending(ValidationStatus::Reattempting(previous_attempts)) => {
+                Some(TxStatus::Pending(ValidationStatus::Reattempting(
+                    previous_attempts.clone().add_attempt(error, clock),
+                )))
+            }
+            TxStatus::Confirmed { .. } => None,
+        }
+    }
+}
+
+impl UpdatableValidationStatus for FailureStatus {
+    fn update_after_failure(
+        &self,
+        error: BlockchainErrorKind,
+        clock: &dyn ValidationFailureClock,
+    ) -> Option<Self> {
+        match self {
+            FailureStatus::RecheckRequired(ValidationStatus::Waiting) => {
+                Some(FailureStatus::RecheckRequired(
+                    ValidationStatus::Reattempting(PreviousAttempts::new(error, clock)),
+                ))
+            }
+            FailureStatus::RecheckRequired(ValidationStatus::Reattempting(previous_attempts)) => {
+                Some(FailureStatus::RecheckRequired(
+                    ValidationStatus::Reattempting(
+                        previous_attempts.clone().add_attempt(error.into(), clock),
+                    ),
+                ))
+            }
+            FailureStatus::RetryRequired | FailureStatus::Concluded => None,
+        }
+    }
+}
+
+pub struct MismatchReport {
+    pub noticed_at: TxHashByTable,
+    pub remaining_hashes: Vec<TxHashByTable>,
+}
+
+pub trait PendingPayableCache<Record> {
```

Comment:
In my opinion, using an enum over generic is better here. Why? A person can easily see what kind of record we're talking about. As of now, they would have to scan through the code to figure out that info. Let's eliminate this inconvenince here.

Reply:
Nope. 

With generics, you can tie the specific implementor up with the specific type replacing the generic. There is no ambiguity then, to one implementor there is only one possible type of record attached. But the enum allows all those possible variants. That's loose boundary which is less favorable than using generics, providing stricter implementation and so better clarity. 

Therefore it is the opposite what you were getting at initially